DevAcademy/Observables/PlacesObservable.swift

```diff
@@ -13,27 +13,66 @@ import SwiftUI
 
 class PlacesObservable: ObservableObject {
     @Published var places: [Place] = []
-    @Published var favorites: [Place] = []
     private let service: PlacesService
 
     init(with service: PlacesService) {
         self.service = service
     }
 
+    var rawPlaces: [Place] = [] {
+        didSet {
+            updatePlaces()
+        }
+    }
+
+    private(set) var favorites: [Int]? {
+        get { (UserDefaults.standard.array(forKey: "favourites") ?? []) as? [Int] }
+        set {
+            UserDefaults.standard.set(newValue, forKey: "favourites")
+            updatePlaces()
+        }
+    }
+
     @MainActor
     func fetchPlaces() async {
         do {
             let result = try await service.places()
-            self.places = result.places
+            self.rawPlaces = result.places
         } catch {
         }
     }
 
-    func addToFavorites(place: Place) {
-        if favorites.contains(place) {
-            favorites.removeAll(where: {$0.properties.ogcFid == place.properties.ogcFid})
+    func isFavorited(place: Place) -> Bool {
+        if let fav = favorites {
+            if fav.contains(where: {$0 == place.properties.ogcFid}) {
+                return true
+            }
+        }
+        return false
+    }
+
+    func setFavorite(place: Place, value: Bool) {
+        if value {
+            if !isFavorited(place: place) {
+                // Add to favorites if not added yet
+                favorites?.append(place.properties.ogcFid)
+            }
         } else {
-            favorites.append(place)
+            // Remove all occurences
+            favorites?.removeAll(where: {$0 == place.properties.ogcFid})
+        }
+    }
+
+    func updatePlaces() {
+        places.append(contentsOf: rawPlaces)
+
+        if let fav = favorites {
+            for id in fav {
+                if let index = places.firstIndex(where: {$0.properties.ogcFid == id}) {
+                    let element = places.remove(at: index)
+                    places.insert(element, at: 0)
+                }
+            }
         }
```

Comment:
```suggestion
        var newPlaces = rawPlaces

        if let fav = favorites {
            for id in fav {
                if let index = newPlaces.firstIndex(where: {$0.properties.ogcFid == id}) {
                    let element = newPlaces.remove(at: index)
                    newPlaces.insert(element, at: 0)
                }
            }
            
            self.places = newPlaces
        }
```

První věc, doporučuji pokud aktualizujete `@Published` proměnnou - vždycky si udělejte kopii a potom tuto proměnnou přepište. Vyhnete se opakovanému generování zbytečných eventů.

Druhý problém je, že vy vlastně do places neustále přidáváte nová místa - tím pádem se vám stejné místo v tabulce objeví tolikrát, kolikrát zavoláte funkci `updatePlaces` :)